extjson.go

```diff
@@ -0,0 +1,1694 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// convertV2Binary is called after the opening quote of the base64 payload. v1
+// $binary is a string, followed by "$type" and no other keys.
+func (d *Decoder) convertV1Binary(out []byte) ([]byte, error) {
+	// Write a length placeholder and a subtype byte placeholder
+	lengthPos := len(out)
+	out = append(out, emptyLength...)
+	subTypeBytePos := len(out)
+	out = append(out, emptyType)
+
+	// Read the payload
+	out, err := d.convertBase64(out)
+	if err != nil {
+		return nil, err
+	}
+
+	// $type key must be next
+	err = d.readCharAfterWS(',')
+	if err != nil {
+		return nil, err
+	}
+	err = d.readQuoteStart()
+	if err != nil {
+		return nil, err
+	}
+	key, err := d.peekBoundedQuote(6, 6)
+	if err != nil {
+		return nil, err
+	}
+	if !bytes.Equal(key, jsonType) {
+		return nil, d.parseError(key[0], "expected $type")
+	}
+	_, _ = d.json.Discard(len(key) + 1)
+	err = d.readNameSeparator()
+	if err != nil {
+		return nil, err
+	}
+	err = d.readQuoteStart()
+	if err != nil {
+		return nil, err
+	}
+	out, err = d.convertBinarySubType(out, subTypeBytePos)
+	if err != nil {
+		return nil, err
+	}
+
+	// write length of binary payload (added length of the output minux 5 bytes
+	// for length+type)
+	binLength := len(out) - lengthPos - 5
+	overwriteLength(out, lengthPos, binLength)
+
+	return out, nil
+}
+
+// convertMinMaxKey starts after the `"$minKey"` or `"$maxKey"` key.  In either
+// case the type byte is already set and the only value for the key is `1` and
+// no futher data has to be written. This function only validates and consumes
+// input.
+func (d *Decoder) convertMinMaxKey(out []byte) ([]byte, error) {
+	// consume ':'
+	err := d.readNameSeparator()
+	if err != nil {
+		return nil, err
+	}
+
+	// Rest must be `1` followed by object terminator
+	err = d.readCharAfterWS('1')
+	if err != nil {
+		return nil, err
+	}
+	err = d.readObjectTerminator()
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// convertSymbol starts after the `"$symbol"` key.
+func (d *Decoder) convertSymbol(out []byte) ([]byte, error) {
+	// consume ':'
+	err := d.readNameSeparator()
+	if err != nil {
+		return nil, err
+	}
+
+	// Must have `"` followed by string followed by object terminator
+	err = d.readCharAfterWS('"')
+	if err != nil {
+		return nil, err
+	}
+	out, err = d.convertString(out)
+	if err != nil {
+		return nil, err
+	}
+	err = d.readObjectTerminator()
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// convertOptions starts after the opening quote of `"$regex"`.  We need to
+// distinguish between Extended JSON $regex or MongoDB $regex query operator.
+// If we can peek far enough, we can check with regular expresssions.
+//
+// See convertRegex for the logic differentiating the query and extended JSON
+// forms.  Unlike that function, the regular expressions here must look past
+// $option to find $regex to disambiguate.
```

Comment:
[nit] move this comment to the line before the element it describes (`convertOptions`)